Make `pipeline` able to load `processor`

qubvel opened the pull request:
# What does this PR do?

Add `processor` to the `pipeline` and refactor tests a bit to support it. Related to 
 - #32490
 - #32059 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

qubvel committed:
Add processor to the pipeline

qubvel committed:
Refactor get_test_pipeline

qubvel committed:
Fixup

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32514). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel committed:
Fixing tests

qubvel committed:
Add processor loading in tests

qubvel committed:
Restructure processors loading

qubvel committed:
Move model loading on tom of the test

qubvel committed:
Update `get_test_pipeline`

qubvel committed:
Fixup

qubvel committed:
Add class-based flags for loading processors

qubvel committed:
Change `is_pipeline_test_to_skip` signature

qubvel committed:
Skip t5 failing test for slow tokenizer

qubvel committed:
Fixup

qubvel committed:
Fix copies for T5

qubvel committed:
Fix typo

qubvel committed:
Add try/except for tokenizer loading (kosmos-2 case)

qubvel committed:
Fixup

qubvel committed:
Llama not fails for long generation

qubvel commented on src/transformers/pipelines/__init__.py:
```diff
@@ -904,9 +926,17 @@ def pipeline(
     load_image_processor = type(model_config) in IMAGE_PROCESSOR_MAPPING or image_processor is not None
+    load_processor = type(model_config) in PROCESSOR_MAPPING or processor is not None
+
+    # Check that pipeline class required loading
+    load_tokenizer = load_tokenizer and pipeline_class._load_tokenizer
+    load_feature_extractor = load_feature_extractor and pipeline_class._load_feature_extractor
+    load_image_processor = load_image_processor and pipeline_class._load_image_processor
+    load_processor = load_processor and pipeline_class._load_processor
```
For backward compatibility, we can control with `Pipeline` class if we need to load specific processors/tokenizers. For example, for zero-shot object detection, we will need to load only the `processor`, and do not need to load `image_processor` and tokenizer separately. Other legacy pipelines might load tokenizer and image_processor, even if they have processor class.

qubvel commented on src/transformers/pipelines/base.py:
```diff
@@ -803,6 +814,18 @@ class Pipeline(_ScikitCompat, PushToHubMixin):
     constructor argument. If set to `True`, the output will be stored in the pickle format.
     """
 
+    # Previously, pipelines support only `tokenizer`, `feature_extractor`, and `image_processor`.
+    # As we start adding `processor`, we want to avoid loading processor for some pipelines, that don't required it,
+    # because, for example, use `image_processor` and `tokenizer` separately.
+    # However, we want to enable it for new pipelines. Moreover, this allow us to granularly control loading components
+    # and avoid loading tokenizer/image_processor/feature_extractor twice: once as a separate object
+    # and once in the processor. The following flags a set this way for backward compatibility ans might be overridden
+    # in specific Pipeline class.
+    _load_processor = False
+    _load_image_processor = True
+    _load_feature_extractor = True
+    _load_tokenizer = True
```
granular control for loading, see comment in the code

qubvel commented on tests/models/mt5/test_modeling_mt5.py:
```diff
@@ -1055,6 +1062,26 @@ def test_with_token_classification_head(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_with_token_classification_head(*config_and_inputs)
 
+    def is_pipeline_test_to_skip(
+        self,
+        pipeline_test_case_name,
+        config_class,
+        model_architecture,
+        tokenizer_name,
+        image_processor_name,
+        feature_extractor_name,
+        processor_name,
+    ):
+        if tokenizer_name is None:
+            return True
+
+        # `MT5EncoderOnlyModelTest` is not working well with slow tokenizers (for some models) and we don't want to touch the file
+        # `src/transformers/data/processors/squad.py` (where this test fails for this model)
+        if pipeline_test_case_name == "TokenClassificationPipelineTests" and not tokenizer_name.endswith("Fast"):
+            return True
+
+        return False
+
```
This one is skipped on the main but was not skipped on this branch, so I added a skip rule here

qubvel replied:
Yes, this test was not running on the main branch. One of the problem of current tests: we are testing a set of cases, and if one of the cases (e.g. the first one) falls under skipTest(...) the rest test cases are not running. Probably it's better to return case status, and then aggregate statuses in the main test function running on multiple cases.

qubvel replied:
A simple example to reproduce:

```python
import unittest

class TestCases(unittest.TestCase):

    def test_multiple_cases(self):
        for i in range(10):
            self.run_test(i)

    def run_test(self, i):
        if i == 0:
            self.skipTest("Skip this test")
        elif i == 5:
            raise ValueError("This test failed")
```

qubvel committed:
Revert processor pass in text-generation test

qubvel commented on tests/pipelines/test_pipelines_text_generation.py:
```diff
@@ -413,6 +431,7 @@ def run_pipeline_test(self, text_generator, _):
             "XGLMForCausalLM",
             "GPTNeoXForCausalLM",
             "FuyuForCausalLM",
+            "LlamaForCausalLM",
```
Llama no longer raises exception for long generation, so I added it to exclude list in the test. This test was skipped on main.

qubvel commented on tests/test_pipeline_mixin.py:
```diff
@@ -144,40 +148,68 @@ def run_task_tests(self, task, torch_dtype="float32"):
 
-            tokenizer_names = []
-            processor_names = []
+            # tokenizers are mapped to tuple, e.g. ("XxxTokenizer", None)
+            tokenizer_names = TOKENIZER_MAPPING_NAMES.get(model_type, [None])
+            if len(tokenizer_names) > 1:
+                tokenizer_names = [name for name in tokenizer_names if name is not None]
+
+            # image processors are mapped to tuple, e.g. ("XxxImageProcessor", None)
+            image_processor_names = IMAGE_PROCESSOR_MAPPING_NAMES.get(model_type, [None])
+            if len(image_processor_names) > 1:
+                image_processor_names = [name for name in image_processor_names if name is not None]
+
+            # feature extractors are mapped to instance, e.g. "XxxFeatureExtractor"
+            feature_extractor_names = FEATURE_EXTRACTOR_MAPPING_NAMES.get(model_type, [None])
+            if not isinstance(feature_extractor_names, (list, tuple)):
+                feature_extractor_names = [feature_extractor_names]
+
+            # processors are mapped to instance, e.g. "XxxProcessor"
+            processor_names = PROCESSOR_MAPPING_NAMES.get(model_type, [None])
+            if not isinstance(processor_names, (list, tuple)):
+                processor_names = [processor_names]
+
```
Previously, tokenizer, image processor and feature extractor class names were taken from JSON file of tiny models, however, there is no `*Processor` class there. Now MAPPING lists are used to get class names.

qubvel replied:
I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

qubvel committed:
Fix docs

qubvel commented on tests/test_pipeline_mixin.py:
```diff
@@ -236,67 +307,89 @@ def run_pipeline_test(
 
+        # -------------------- Load model --------------------
+
+        # TODO: We should check if a model file is on the Hub repo. instead.
+        try:
+            model = model_architecture.from_pretrained(repo_id, revision=commit)
+        except Exception:
+            logger.warning(
+                f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')}_{torch_dtype} is skipped: Could not find or load "
+                f"the model from `{repo_id}` with `{model_architecture}`."
+            )
+            self.skipTest(f"Could not find or load the model from {repo_id} with {model_architecture}.")
+
+        # -------------------- Load tokenizer --------------------
+
```
The order of loading is a bit changed in this test, now the model is loaded first to check if it exists, otherwise, the test is skipped.

qubvel reviewed (commented):
The main changes for pipeline code:
 - Added loading of `processor` (previously only tokenizer, image_processor and feature_extractor)
 - Added Pipeline class args to control loading
 
 The main changes 
 - Refactored test signatures to support `processor` pass
 - Changed the way how class names are obtained for processors of the testing tiny models in a pipeline. Instead of getting them from JSON file, the MAPPINGs are used.
 
 I highlighted these changes below, please see the comments
 
cc @ydshieh for initial review

qubvel commented:
@ydshieh yes, you are right, the [following PR](https://github.com/huggingface/transformers/pull/32490) will add processor to the ZeroShotObjectDetection pipeline

ydshieh commented:
Hi @qubvel I have a general question. The process class could load also a single tokenizer, image processor or feature extractor (if not all components are on a Hub repository). For example,

```python
from transformers import AutoProcessor
AutoProcessor.from_pretrained("bert-base-uncased")
```
will turn out to be a tokenizer instead of an instance of `ProcessorMixin`.

Therefore, in `src/transformers/pipelines/__init__.py`, this part

```python
       # Instantiate processor if needed
        if isinstance(processor, (str, tuple)):
            processor = AutoProcessor.from_pretrained(processor, _from_pipeline=task, **hub_kwargs, **model_kwargs)
```
should have some guard to make sure we really get a `ProcessorMixin` instance.

WDYT?

I will share my thoughts regarding the testing part this afternoon, somehow related to the same consideration.

ydshieh commented on tests/test_pipeline_mixin.py:
```diff
@@ -236,67 +311,89 @@ def run_pipeline_test(
+                try:
+                    # Can fail if some extra dependencies are not installed
+                    processor_class = getattr(transformers_module, name)
+                    processor = processor_class.from_pretrained(repo_id, revision=commit)
```
we should make sure `processor` is an instance of `ProcessorMixin`, otherwise reset it to `None`

qubvel replied:
should we? Class is not checked for feature extractor or image processor, I guess we might want to load anything that might be loaded with AutoProcessor class while it keeps the pipeline working.

ydshieh replied:
> Class is not checked for feature extractor or image processor

This is because they are not a composite, and are not design to fallback to load partially.

For example, this works
```
from transformers import AutoProcessor
AutoProcessor.from_pretrained("bert-base-uncased")
```
but this fails
```
from transformers import AutoImageProcessor
AutoImageProcessor.from_pretrained("bert-base-uncased")
```
and same
```
from transformers import AutoTokenizer
AutoTokenizer.from_pretrained("google/vit-base-patch16-224")
```

qubvel replied:
Got it, addressed in https://github.com/huggingface/transformers/pull/32514/commits/2f561479290970fa8beaf668b504df39a2608f4f

ydshieh reviewed (commented):
Some comments :-)

qubvel commented:
<img width="886" alt="Screenshot 2024-08-14 at 16 51 57" src="https://github.com/user-attachments/assets/cbe78b28-f39a-4086-8f3a-e98fe098961f">
do not have a `Reply...` field regarding MAPPING (github bug?)

I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

qubvel commented on tests/test_pipeline_mixin.py:
```diff
@@ -236,67 +307,89 @@ def run_pipeline_test(
 
         # validate
-        validate_test_components(self, task, model, tokenizer, processor)
+        validate_test_components(model, tokenizer)
```
can you provide more details, please? this function is used only here and not using image_processor or feature_extractor (I updated the function signature accordingly)

ydshieh replied:
Yeah, it's more a design instead of what is actually happening so far.

The picture is that, a pipeline test is involved pipeline class (task), model, tokenizer/image processor/feature extractor/processor, and all of these **could** determine what we want to perform. (although currently only model and tokenizer appears in the body)

qubvel replied:
Thanks for the clarification, I understand your point. 
While this is a design decision, I'm not entirely comfortable with it. The function isn't widely used or overridden in subclasses, it's only used once in the codebase. Not using arguments might cause confusion about what's actually happening in the function and could give the false impression that the image processor and tokenizer are being validated by it, but they are not. We can always add additional arguments to the function if they become necessary.

ydshieh replied:
works for me.

(FYI: unused arguments exist in `class Pipeline.__init__` and I agree with your above point)

ydshieh commented:
> I did it because in the original tiny-models JSON file we do not have separate lists for image processors and feature extractors, I found they are stored in the same list (even processors are sometimes in this list).

Yes, you are correct. It is because I treated them (image processor / feature exactor) equally, and later I decided to make `processor` join into this pipeline testing

I see better now why you need this change. Could you, at least for now, simply use the information from the summary file `processor_classes` field, and by looking at training part to decide to assign them to one of the  `tokenizer_names`, `feature_extractor_names`, `processor_names` etc.?
``

qubvel committed:
Switch back to json file for image processors and feature extractors

qubvel committed:
Add processor type check

qubvel commented on tests/test_pipeline_mixin.py:
```diff
@@ -236,67 +307,89 @@ def run_pipeline_test(
+                        f"{self.__class__.__name__}::test_pipeline_{task.replace('-', '_')}_{torch_dtype} is skipped: "
+                        f"Could not load the {key} from `{repo_id}` with `{name}`."
+                    )
+                    self.skipTest(f"Could not load the {key} from {repo_id} with {name}.")
```
It is designed the same way on the `main` now, according to the comment some instances might not be loaded due to optional dependencies
https://github.com/huggingface/transformers/blob/af638c4afea81ac251f1ae63497789be79e0baaa/tests/test_pipeline_mixin.py#L256-L267

ydshieh replied:
Currently on `main`, this `run_pipeline_test` only take a single `tokenizer_name` and a **single** `processor_name`, and try to load them. So if a specified `processor_name` could not be loaded, we can skip it.

However, here in this PR, we try to **collect** the processors and put them into the dictionary `processors`. So logically it's better not to skip the test during this collection.

(But in practice, this may not produce any difference)

qubvel committed:
Remove except for tokenizers

qubvel commented:
@ydshieh Thanks for reviewing, I addressed the comments, please have a look!

> Once we start to have the code using processor, and if we support it for some currently existing pipelines, it should support:
> - either not use the processor at all (i.e. currently implementation): for backward compatibility
> - or completely use the processor only (but need to verify the results are the same)

This is what I tried to achieve with granular control for each pipeline class by adding `_load_*` attributes 
https://github.com/huggingface/transformers/pull/32514#discussion_r1711361484

qubvel commented:
@ydshieh could you please have a look once again?

ydshieh commented on tests/test_pipeline_mixin.py:
```diff
@@ -236,67 +324,82 @@ def run_pipeline_test(
             tokenizer_name (`str`):
                 The name of a subclass of `PreTrainedTokenizerFast` or `PreTrainedTokenizer`.
-            processor_name (`str`):
+            image_processor_or_feature_extractor_name (`str`):
```
This looks like not the correct docstring?

qubvel replied:
addressed in https://github.com/huggingface/transformers/pull/32514/commits/403cd0b810413da5e462f28410c224e9d8eca39d

ydshieh commented on tests/test_pipeline_mixin.py:
```diff
@@ -144,40 +145,84 @@ def run_task_tests(self, task, torch_dtype="float32"):
-            # Adding `None` (if empty) so we can generate tests
-            tokenizer_names = [None] if len(tokenizer_names) == 0 else tokenizer_names
-            processor_names = [None] if len(processor_names) == 0 else processor_names
+            if model_arch_name in tiny_model_summary and "sha" in tiny_model_summary[model_arch_name]:
```
the first condition is not necessary at this point.

qubvel replied:
otherwise we can get an error while doing `tiny_model_summary[model_arch_name]`

ydshieh commented on tests/test_pipeline_mixin.py:
```diff
@@ -144,40 +145,84 @@ def run_task_tests(self, task, torch_dtype="float32"):
+            if model_arch_name not in tiny_model_summary:
+                continue
+
+            tokenizer_names = tiny_model_summary[model_arch_name]["tokenizer_classes"]
+
+            # Sort image processors and feature extractors from tiny-models json file
+            image_processor_names = []
+            feature_extractor_names = []
```
if we don't add `None` later in this method (while it's empty), then in `run_model_pipeline_tests`, 

> test_cases = [ ....]

will be empty if `tokenizer_names` is empty or `image_processor_names` is empty or ... (etc).

But we should allow the test case to exist even if only one of them present: for example, a text only model or a vision only model.

ydshieh reviewed (commented):
Just a few nits but overall good!

Let me know if my comments are clear 🙏

yonigozlan commented:
Hi @qubvel @ydshieh ! Just wondering if anything is blocking this PR from getting merged, as it is needed for the image-text-to-text pipeline. Thanks!

qubvel commented:
Hi @yonigozlan, just a lack of time 🙂 I will address comments this week, and hopefully it can be merged then!

qubvel committed:
Fix docstring

qubvel committed:
Fix empty lists for tests

HuggingFaceDocBuilderDev commented:
Hey! 🤗 Thanks for your contribution to the `transformers` library!

Before merging this pull request, slow tests CI should be triggered. To enable this:
- Add the `run-slow` label to the PR
- When your PR is ready for merge and all reviewers' comments have been addressed, push an empty commit with the command `[run-slow]` followed by a comma separated list of all the models to be tested, i.e. `[run_slow] model_to_test_1, model_to_test_2`
  - If the pull request affects a lot of models, put at most 10 models in the commit message
- A `transformers` maintainer will then approve the workflow to start the tests

(For maintainers) The documentation for slow tests CI on PRs is [here](https://www.notion.so/huggingface2/CI-for-pull-requests-8335bd9217d24d1e8ed1d3a016f39804).

qubvel committed:
Fixup

qubvel commented:
@Rocketknight1 can you please review the PR, if that aligns with plans for `pipeline` refactoring/standardization

Rocketknight1 commented:
I'm sorry, I missed that ping somehow! Reviewing now.

Rocketknight1 commented on src/transformers/pipelines/__init__.py:
```diff
@@ -905,13 +927,29 @@ def pipeline(
-    load_image_processor = type(model_config) in IMAGE_PROCESSOR_MAPPING or image_processor is not None
+    load_feature_extractor = (
+        type(model_config) in FEATURE_EXTRACTOR_MAPPING
+        or feature_extractor is not None
+        or isinstance(feature_extractor, str)
```
I think this conditional is redundant - if `feature_extractor` is a `str` then it will always be `not None` as well.

qubvel replied:
addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

Rocketknight1 commented on src/transformers/pipelines/__init__.py:
```diff
@@ -905,13 +927,29 @@ def pipeline(
+    )
+    load_image_processor = (
+        type(model_config) in IMAGE_PROCESSOR_MAPPING
+        or image_processor is not None
+        or isinstance(image_processor, str)
```
Same here!

qubvel replied:
addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

Rocketknight1 commented on src/transformers/pipelines/__init__.py:
```diff
@@ -905,13 +927,29 @@ def pipeline(
+        or image_processor is not None
+        or isinstance(image_processor, str)
+    )
+    load_processor = type(model_config) in PROCESSOR_MAPPING or processor is not None or isinstance(processor, str)
```
And here!

qubvel replied:
addressed in https://github.com/huggingface/transformers/pull/32514/commits/d553dab8a33c4144b29c649115fb3a41cd92d6bf

Rocketknight1 commented on tests/test_pipeline_mixin.py:
```diff
@@ -218,27 +271,66 @@ def run_model_pipeline_tests(
-                    continue
-                self.run_pipeline_test(
-                    task, repo_name, model_architecture, tokenizer_name, processor_name, commit, torch_dtype
+        test_cases = [
+            {
+                "tokenizer_name": tokenizer_name,
+                "image_processor_name": image_processor_name,
+                "feature_extractor_name": feature_extractor_name,
+                "processor_name": processor_name,
+            }
+            for tokenizer_name in tokenizer_names
+            for image_processor_name in image_processor_names
+            for feature_extractor_name in feature_extractor_names
+            for processor_name in processor_names
+        ]
```
Note for other reviewers: When a model doesn't support one of these preprocessor types, the list of names will be `[None]`, **not** an empty list. When I first looked at this quadruple-loop I was worried that `test_cases` would be empty if any of the four name lists was empty, but that should never occur. For safety, we might want to add a check for that, though, since it depends on behaviour in a different function that might be modified later!

qubvel replied:
You are right, let's have this check closer to the loop to ensure we always have something to test
https://github.com/huggingface/transformers/pull/32514/commits/6ff4e68ec9caa17f4f5f2535f00f0fbca5fda115

ydshieh replied:
After this change, you can also remove

```
            if not isinstance(processor_names, (list, tuple)):
                processor_names = [processor_names]
```
(which was there for the same reason)

qubvel replied:
I left it there to make sure we do not have `str` processor_names

Rocketknight1 reviewed (approved):
Just finished the review, with a few comments.

Overall summary for core maintainers:
- This is a nice PR! It's clean, and although it seems very large, almost all of the changes are boilerplate adding kwargs and fixing spelling mistakes in a lot of files.
- The actual changes you need to look at are `__init__.py`, `base.py` and `test_pipeline_mixin.py`
- I made some comments, but they're nits and shouldn't affect any of the PR logic.

cc @lysandrejik for core maintainer review!

qubvel committed:
Fix load check

qubvel committed:
Ensure we have non-empty test cases

LysandreJik reviewed (commented):
This looks good! I'm mostly concerned about having good and clear documentation and removing unnecessary warnings.

Pipelines are the very first transformers object many users interact with -> ensuring that they're aware of the info they need, without being hammered by warnigns should be an absolute goal of ours.

Thanks for working on this important change!

qubvel committed:
Update src/transformers/pipelines/__init__.py

Co-authored-by: Lysandre Debut <hi@lysand.re>

qubvel committed:
Update src/transformers/pipelines/base.py

Co-authored-by: Lysandre Debut <hi@lysand.re>

qubvel committed:
Rework comment

qubvel committed:
Better docs, add note about pipeline components

qubvel commented on src/transformers/pipelines/__init__.py:
```diff
@@ -644,6 +647,25 @@ def pipeline(
             is a string). However, if `config` is also not given or not a string, then the default feature extractor
             for the given `task` will be loaded.
+        image_procesor (`str` or [`BaseImageProcessor`], *optional*):
+            The image processor that will be used by the pipeline to preprocess images for the model. This can be a
+            model identifier or an actual image processor inheriting from [`BaseImageProcessor`].
+
+            Image processors are used for Vision models and multi-modal models that require image inputs. Multi-modal
+            models will also require a tokenizer to be passed.
+
+            If not provided, the default image processor for the given `model` will be loaded (if it is a string). If
+            `model` is not specified or not a string, then the default image processor for `config` is loaded (if it is
+            a string).
+        processor (`str` or [`ProcessorMixin`], *optional*):
+            The processor that will be used by the pipeline to preprocess data for the model. This can be a model
+            identifier or an actual processor inheriting from [`ProcessorMixin`].
+
+            Processors are used for multi-modal models that require multi-modal inputs, for example, a model that
+            requires both text and image inputs.
+
+            If not provided, the default processor for the given `model` will be loaded (if it is a string). If `model`
+            is not specified or not a string, then the default processor for `config` is loaded (if it is a string).
```
I added a separate Note section to highlight that we should not provide all types of `processors` at once
https://github.com/huggingface/transformers/pull/32514/commits/e7127176ae83dd401a0e0358aa6cecd5db80fe9b and refer to a specific pipeline in case one would like to provide them explicitly.

For each specific pipeline we have only required processors args in docs section configured with docs decorator. e.g. here
https://github.com/huggingface/transformers/blob/e782e95e3465b66a377c0a7fe95f8f10cd87459d/src/transformers/pipelines/image_classification.py#L53

qubvel replied:
Updated in https://github.com/huggingface/transformers/pull/32514/commits/699669503629a834013d9c2107396051b3b9327e

qubvel commented on src/transformers/pipelines/base.py:
```diff
@@ -805,6 +816,18 @@ class Pipeline(_ScikitCompat, PushToHubMixin):
 
+    # Previously, pipelines support only `tokenizer`, `feature_extractor`, and `image_processor`.
+    # As we start adding `processor`, we want to avoid loading processor for some pipelines, that don't required it,
+    # because, for example, use `image_processor` and `tokenizer` separately.
```
Tried to make it more clear https://github.com/huggingface/transformers/pull/32514/commits/9793e012a62dfda59942b586e0e21b612b5e0cfb

qubvel committed:
Change warning to error raise

qubvel committed:
Fixup

qubvel commented on src/transformers/pipelines/__init__.py:
```diff
@@ -1074,6 +1100,32 @@ def pipeline(
+                    f"Processor will be not loaded, because {processor} is not an instance of `ProcessorMixin`. "
+                    f"Got type `{type(processor)}` instead.",
+                    UserWarning,
+                )
```
Thanks for the questions!  
There was a discussion [here](https://github.com/huggingface/transformers/pull/32514#discussion_r1717215386). The main idea is that `AutoProcessor` can load not only processors but also basic processing classes, like `tokenizer`. 

Indeed, a misconfiguration in the model-pipeline-processor setup could trigger this and raising a warning + dropping the `processor` might be sufficient **only when the `processor` isn't needed in the `pipeline` at all.** 

However, with granular control, such a case shouldn't occur. Therefore, it seems more appropriate to replace the warning with an error to clearly indicate a misconfiguration. Otherwise, the error will happen later with a less clear message because the `processor` is None.

qubvel replied:
Changed to error raise here https://github.com/huggingface/transformers/pull/32514/commits/bcce4dceb19b39225b11acbcc495a754fa6c802d

qubvel committed:
Merge branch 'main' into add-processor-to-pipeline

qubvel committed:
Refine pipeline docs

qubvel commented:
@LysandreJik thanks for reviewing, I addressed the comment: made clearer docs and removed the warning. Please have a look whenever you have bandwidth.

P.S. The test failure looks unrelated.

LysandreJik reviewed (approved):
Ok great, thanks for the changes @qubvel!

qubvel committed:
Merge branch 'main' into add-processor-to-pipeline
